### PR TITLE
Add validation for LAG port update request

### LINKED ISSUE #4719: Add validation for LAG port update request
There is API `/v2/switches/{switch_id}/lags/{logical_port_number}` which allows to add/remove physical ports into LAG group.

We should  allow to decrease the LAG port capacity only to the consumed BW value. So, if for example, we have two ports in a LAG group  (10G each) with 20G total BW and the consumed is 18G, we should not allow to remove the port from the group. 
If the consumed BW is less than 10 G we should allow to remove the port from the lag group

In case of error Kilda must return 400 Bad request response.

### PR DESCRIPTION
**Pull request by vshakirova**
Close #4719

### CODE REVIEW

**Review by surabujin — CHANGES_REQUESTED** [review]
One area related to port speed is missing/not changed - ISL speed. We need to have at least note/todo record about changing ISL speed on OTHER_UPDATE event. Mentioned chage for ISL speed can become to complex/big for current PR.

File: docker/db-migration/migrations/020-add-port-class.yaml [hunk 1]
```diff
@@ -0,0 +1,21 @@
+databaseChangeLog:
+  - changeSet:
+      id: tag
+      author: vshakirova
+      changes:
+        - tagDatabase:
+            tag: 020-add-port-class
+
+  - changeSet:
```
[the remark this answers is not in the document]
Comment (vshakirova) [reply docker/db-migration/migrations/020-add-port-class.yaml:9]: Yep, now we have relation between port and switch obj. But it is described in FermaPortRepo. Not sure about related class, what do you mean?

File: docs/design/network-discovery/switch-FSM.puml [hunk 2]
```diff
@@ -58,19 +59,25 @@ SYNC -r-> OFFLINE : offline
 SETUP : enter / update status in DB
 SETUP : enter / emit switch online status notification
 SETUP : enter / setup port FSM for NEW ports
+SETUP : enter / save NEW port in BD
```
[the remark this answers is not in the document]
Comment (vshakirova) [reply docs/design/network-discovery/switch-FSM.puml:62]: yes, sorry. my mistake. I'll fix this issue.

File: docker/db-migration/migrations/020-add-port-class.yaml [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+      id: add_port_class
+      author: vshakirova
+      changes:
+        - sql: "CREATE CLASS port IF NOT EXISTS EXTENDS V"
+        - sql: "CREATE PROPERTY port.switch_id IF NOT EXISTS STRING"
+        - sql: "CREATE PROPERTY port.port_no IF NOT EXISTS INTEGER"
+        - sql: "CREATE PROPERTY port.curr_speed IF NOT EXISTS LONG"
+        - sql: "CREATE PROPERTY port.max_speed IF NOT EXISTS LONG"
+        - sql: "CREATE INDEX port.switch_id IF NOT EXISTS NOTUNIQUE_HASH_INDEX"
```
[the remark this answers is not in the document]
Comment (vshakirova) [reply docker/db-migration/migrations/020-add-port-class.yaml:18]: I will do that. thank you

File: src-java/network-topology/network-storm-topology/src/main/java/org/openkilda/wfm/topology/network/controller/sw/PhysicalPort.java [hunk 4]
```diff
@@ -28,14 +29,29 @@
 @EqualsAndHashCode(callSuper = true)
 public class PhysicalPort extends AbstractPort {
     private Isl history;
+    private long maxSpeed;
+    private long currSpeed;
```
[the remark this answers is not in the document]
Comment (vshakirova) [reply src-java/network-topology/network-storm-topology/src/main/java/org/openkilda/wfm/topology/network/controller/sw/PhysicalPort.java:33]: oki, will fix that)

File: src-java/network-topology/network-storm-topology/src/main/java/org/openkilda/wfm/topology/network/controller/sw/SwitchFsm.java [hunk 5]
```diff
@@ -298,7 +308,8 @@ public void handlePortLinkStateChange(SwitchFsmState from, SwitchFsmState to, Sw
                 throw new IllegalArgumentException(String.format("Unexpected event %s received in state %s (%s)",
                         event, to, getClass().getName()));
         }
-
+        portUpdate(makePortRecord(Endpoint.of(switchId, context.getPortNumber()), context.getPortMaxSpeed(),
```
[the remark this answers is not in the document]
Comment (vshakirova) [reply src-java/network-topology/network-storm-topology/src/main/java/org/openkilda/wfm/topology/network/controller/sw/SwitchFsm.java:311]: fixed

File: src-java/kilda-persistence-tinkerpop/src/main/java/org/openkilda/persistence/ferma/frames/PortFrame.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+import com.syncleus.ferma.annotations.Property;
+import org.apache.tinkerpop.gremlin.structure.Direction;
+import org.apache.tinkerpop.gremlin.structure.Edge;
+
+import java.util.List;
+import java.util.Objects;
+
+public abstract class PortFrame extends KildaBaseVertexFrame implements PortData {
+    public static final String FRAME_LABEL = "port";
+    public static final String OWNS_BY_PORT = "owns";
```
[the remark this answers is not in the document]
Comment (vshakirova) [reply src-java/kilda-persistence-tinkerpop/src/main/java/org/openkilda/persistence/ferma/frames/PortFrame.java:35]: Lets try this)

File: src-java/network-topology/network-storm-topology/src/main/java/org/openkilda/wfm/topology/network/controller/port/PortFsm.java [hunk 7]
```diff
@@ -122,21 +143,53 @@ private void emitRoundTripInactive(IPortCarrier carrier) {
...
+                                .build();
+                        portRepository.add(newPort);
+                        return newPort;
+                    });
+            port.setMaxSpeed(portData.getMaxSpeed());
+            port.setCurrentSpeed(portData.getCurrentSpeed());
+        });
+    }
+
+    private void createOrUpdate(PortFsmState from, PortFsmState to, PortFsmEvent event, PortFsmContext context) {
```
[the remark this answers is not in the document]
Comment (vshakirova) [reply src-java/network-topology/network-storm-topology/src/main/java/org/openkilda/wfm/topology/network/controller/port/PortFsm.java:167]: Will move it, thank you

File: src-java/network-topology/network-storm-topology/src/main/java/org/openkilda/wfm/topology/network/controller/port/PortFsm.java [hunk 8]
```diff
@@ -122,21 +153,53 @@ private void emitRoundTripInactive(IPortCarrier carrier) {
...
+        });
+    }
+
     // -- service data types --
 
     public static class PortFsmFactory {
         private final StateMachineBuilder<PortFsm, PortFsmState, PortFsmEvent, PortFsmContext> builder;
-        private final PortPropertiesRepository portPropertiesRepository;
-
-        PortFsmFactory(PersistenceManager persistenceManager) {
```
[the remark this answers is not in the document]
Comment (vshakirova) [reply src-java/network-topology/network-storm-topology/src/main/java/org/openkilda/wfm/topology/network/controller/port/PortFsm.java:131]: thanks, moved it

File: src-java/network-topology/network-storm-topology/src/main/java/org/openkilda/wfm/topology/network/controller/port/PortFsm.java [hunk 9]
```diff
@@ -122,21 +153,53 @@ private void emitRoundTripInactive(IPortCarrier carrier) {
...
+                                        endpoint.getDatapath())));
+                        Port newPort = Port.builder()
+                                .switchObj(sw)
+                                .switchId(new SwitchId(endpoint.getDatapath().getId()))
+                                .portNo(endpoint.getPortNumber())
+                                .build();
+                        portRepository.add(newPort);
+                        return newPort;
+                    });
+            port.setMaxSpeed(portData.getMaxSpeed());
```
[the remark this answers is not in the document]
Comment (vshakirova) [reply src-java/network-topology/network-storm-topology/src/main/java/org/openkilda/wfm/topology/network/controller/port/PortFsm.java:172]: I'll place a simple null check and leave a TODO comment

File: src-java/base-topology/base-storm-topology/src/test/java/org/openkilda/wfm/share/mappers/PortMapperTest.java [hunk 10]
```diff
@@ -32,15 +31,23 @@ public class PortMapperTest {
     @Test
     public void portMapperTest() {
         PortMapper portMapper = Mappers.getMapper(PortMapper.class);
+        PortInfoData portInfoData = new PortInfoData(TEST_SWITCH_ID, 1, 10000000L,
+                10000000L, PortChangeType.UP);
```
Comment (niksv) [inline src-java/base-topology/base-storm-topology/src/test/java/org/openkilda/wfm/share/mappers/PortMapperTest.java:35]: cur speed and max speed still equal

**Review by niksv — APPROVED** [review]

**Review by surabujin — APPROVED** [review]

**Review by surabujin — APPROVED** [review]
LGTM

**Review by dpoltavets — APPROVED** [review]

**Pull request merged**